File: Prim.py
import time
import numpy as np
from Grafos import Grafos

def prim(grafo: Grafos, vertice = 0):
    tempo_inicio = time.time()

    S = []
    Q = list(range(len(grafo.vertices)))
    visitado = np.zeros(len(grafo.vertices), dtype=bool)

    u = vertice
    visitado[u] = True
    Q.remove(u)

    while Q:
        menor = float('inf')
        aresta = (-1, -1)
        for i in range(len(grafo.vertices)):
            if visitado[i]:
                for j in grafo.retornarVizinhos(i):
                    if not visitado[j]:
                        peso = grafo.pesoAresta(i, j)
                        if peso < menor:
                            menor = peso
                            aresta = (i, j)

        if aresta != (-1, -1):
            S.append((aresta[0], aresta[1], menor))
            visitado[aresta[1]] = True
            Q.remove(aresta[1])

    tempo_fim = time.time()

    custo_total = sum([p for _, _, p in S])
    print("AGM:")
    for u, v, p in S:
        print(f"{grafo.labelVertice(u)} - {grafo.labelVertice(v)}: {float(p)}")
    print(f"Custo total: {custo_total}")
    print("Tempo de execução: %.6f segundos\n" % (tempo_fim - tempo_inicio))

File: Java/Grafo/Prim.java
package Grafo;

import java.util.*;

public class Prim {
    public static void executar(Grafo g, int verticeInicial) {
        long inicio = System.nanoTime();

        int n = g.vertices.size();
        boolean[] visitado = new boolean[n];
        List<int[]> arvore = new ArrayList<>();

        visitado[verticeInicial] = true;

        while (true) {
            double menor = Double.POSITIVE_INFINITY;
            int origem = -1;
            int destino = -1;

            for (int i = 0; i < n; i++) {
                if (visitado[i]) {
                    for (int j : g.retornarVizinhos(i)) {
                        if (!visitado[j]) {
                            double peso = g.pesoAresta(i, j);
                            if (peso < menor) {
                                menor = peso;
                                origem = i;
                                destino = j;
                            }
                        }
                    }
                }
            }

            if (origem == -1) break;

            arvore.add(new int[]{origem, destino});
            visitado[destino] = true;
        }

        long fim = System.nanoTime();
        double tempo = (fim - inicio) / 1e9;

        System.out.println("AGM (Prim):");
        for (int[] e : arvore) {
            System.out.printf("%s - %s (%.2f)\n",
                g.labelVertice(e[0]), g.labelVertice(e[1]),
                g.pesoAresta(e[0], e[1]));
        }
        System.out.printf("Tempo de execução: %.6f segundos\n", tempo);
    }
}
